openwrt-openwisp-monitoring/files/monitoring.agent

```diff
@@ -26,76 +29,184 @@ if [ "$VERSION" -eq "1" ]; then
+  if [ "$available" -ge "$required" ]; then
+    echo "1"
+  else
+    [ "$verbose_mode" -eq "1" ] && logger -s "Not enough memory available" \
+                                        -p daemon.err
+    echo "0"
+  fi
+}
 
 collect_data(){
   n=0
-  [ "$verbose_mode" = "1" ] && logger -s "Collecting NetJSON Monitoring data" \
+  [ "$verbose_mode" -eq "1" ] && logger -s "Collecting NetJSON Monitoring data" \
                                       -p daemon.info
   until [ "$n" -ge 5 ]
   do
     echo "$(/usr/sbin/netjson_monitoring "$MONITORED_INTERFACES")" && break
     
     if [ "$n" -eq 5 ]; then
-      [ "$verbose_mode" = "1" ] && logger -s "Collecting data failed!" \
+      [ "$verbose_mode" -eq "1" ] && logger -s "Collecting data failed!" \
                                           -p daemon.err
     fi
     n=$((n+1))
     sleep 5
   done
 }
 
-while true
-do
-  data="$(collect_data)"
-  #send data
-  response_code=$($CURL_COMMAND -H "Content-Type: application/json" \
-    -d "$data" \
-    -v "$URL")
-  if [ "$response_code" = "200" ]; then
-    [ "$verbose_mode" = "1" ] && logger -s "Data sent successfully." \
-                                        -p daemon.info
-  else
-    logger -s "Data not sent successfully. Response code is $response_code" \
+set_url_and_curl(){
+  if [ -z "$BASE_URL" ]; then
+    logger -s "missing required --url option" \
            -t openwisp_monitoring \
            -p daemon.err
+    exit 1
+  fi
 
-    [ "$verbose_mode" = "0" ] && logger -s "Run with verbose mode to find more." \
-                                        -t openwisp_monitoring \
-                                        -p daemon.err
+  if [ -z "$UUID" ]; then
+    logger -s "missing required --uuid option" \
+            -t openwisp_monitoring \
+            -p daemon.err
+    exit 1
   fi
-  sleep "$INTERVAL" & wait $!
-done
+
+  if [ -z "$KEY" ]; then
+    logger -s "missing required --key option" \
+           -t openwisp_monitoring \
+           -p daemon.err
+    exit 1
+  fi
+
+  URL="$BASE_URL/api/v1/monitoring/device/$UUID/?key=$KEY"
+
+  CURL_COMMAND="curl -s -w "%{http_code}""
+  [ "$VERIFY_SSL" -eq "0" ] && CURL_COMMAND="$CURL_COMMAND -k"
+  [ "$verbose_mode" -eq "1" ] && CURL_COMMAND="$CURL_COMMAND -v"
+  MAX_RETRIES=${MAX_RETRIES:-5}
+}
+
+save_data() {
+  while true
+  do
+    memory_available="$(check_available_memory)"
+    if [ "$memory_available" -ne "0" ]; then
+      data="$(collect_data)"
+      # extra zeroes are added for nanoseconds precision
+      file_name="$(date -u +'%d-%m-%Y_%H:%M:%S').000000.json"
+      # make directory 
+      mkdir -p "$TMP_DIR"
+      # save data with file_name 
+      echo "$data" > "$TMP_DIR/$file_name"
+      [ "$verbose_mode" -eq "1" ] && logger -s "Data saved temporarily" \
+                                          -p daemon.info
+    fi
+    # get process id of the process sending data 
+    pid=$(pgrep -f "openwisp_monitoring.*uuid")
```

Comment:
We also don't need to evaluate process id each time but in starting collect data will starts first before send data :thinking:

Reply:
but the other process may be restarted